snuba/datasets/entities/transactions.py

```diff
@@ -85,21 +98,117 @@
 
+        pipeline_builder = SimplePipelineBuilder(
+            query_plan_builder=SingleStorageQueryPlanBuilder(
+                storage=storage,
+                mappers=transaction_translator
+                if custom_mappers is None
+                else transaction_translator.concat(custom_mappers),
+            ),
+        )
+        sampled_pipeline_builder = SampledSimplePipelineBuilder(
+            query_plan_builder=SingleStorageQueryPlanBuilder(
+                storage=storage,
+                mappers=transaction_translator
+                if custom_mappers is None
+                else transaction_translator.concat(custom_mappers),
+            ),
+        )
+
+        def selector_func(query: LogicalQuery, referrer: str) -> Tuple[str, List[str]]:
+            if referrer == "tagstore.__get_tag_keys":
+                condition = query.get_condition()
+                assert condition is not None
+                project_literals = get_literals_for_column_condition(
+                    "project_id", int, condition
+                )
+                values = set(lit.value for lit in project_literals)
+                if 1 in values:
+                    sample_query_rate = state.get_config(
+                        "snuplicator_sampled_query", 0.0
+                    )
+                    assert isinstance(sample_query_rate, float)
+                    if random.random() < sample_query_rate:
+                        return "transactions", ["sampler"]
+
+            return "transactions", []
+
+        def callback_func(
+            query: LogicalQuery,
+            settings: RequestSettings,
+            referrer: str,
+            results: List[Result[QueryResult]],
+        ) -> None:
+            if not results:
+                metrics.increment(
+                    "query_result",
+                    tags={"match": "empty", "primary": "none", "referrer": referrer},
+                )
+                return
+
+            primary_result = results.pop(0)
+            primary_function_id = primary_result.function_id
+            primary_result_data = primary_result.result.result["data"]
+            secondary_result = results.pop(0) if len(results) > 0 else None
+
+            # compare results
+            comparison = "one_result"
+            if secondary_result:
+                secondary_result_data = secondary_result.result.result["data"]
+
+                primary_found_keys = set(row["tags_key"] for row in primary_result_data)
+                secondary_found_keys = set(
+                    row["tags_key"] for row in secondary_result_data
+                )
+
+                if primary_found_keys == secondary_found_keys:
+                    comparison = "match"
+                else:
+                    comparison = "no_match"
+
+            # If we're returning the sampled query we need to multiply the results to ensure they're roughly accurate
+            if primary_function_id == "sampler":
```

Comment:
Was there a reason you included this block? This condition will never be met

Reply:
It will be met if the main query fails, won't it? If the regular query times out but the sampled one doesn't, there will only be one result in the results set and the primary result will be the sampled one.

Reply:
No, the primary query is always the primary query. If it would have previously failed then it should still fail. Snuplicator's only purpose is for running other queries in the background for comparison and experimentation, it will never give a different result to a user than what would be provided by the function marked as "primary".

Reply:
The callback typically happens after the main function is returned, and in any case has no influence on what results the user actually gets. But you're right it is possible for the primary result to be missing from the results list passed to the callback if it failed. It may be a good idea to change the signature of the callback function to more explicitly mark if we did not get a primary result.